### PR TITLE
BinaryData API updates

### LINKED ISSUE #14346: BinaryData - Add static serializer property to avoid unnecessary allocations
We should have something like:
``` private static JsonObjectSerializer DefaultSerializer {get;} = new JsonObjectSerializer(); ```
to avoid unnecessary allocations when using the default set up.

### LINKED ISSUE #14428: Binary Data API Updates based on UX study feedback
- Add FromString
- Add FromBytes(ReadOnlySpan<byte>)
- Add ToBytes method instead of having public Bytes property.
- Add ctor taking ReadOnlyMemory<byte>
- Rename Serialize to FromObject
- Rename Deserialize to ToObject
- Rename FromMemory(byte[]) and FromMemory(ReadOnlyMemory<byte>) to FromBytes
- Make ToString docs more attention grabbing

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/14428 and https://github.com/Azure/azure-sdk-for-net/issues/14346

### CODE REVIEW

File: sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs [hunk 1]
```diff
@@ -5,26 +5,32 @@ namespace Azure.Core
     {
         private readonly object _dummy;
         private readonly int _dummyPrimitive;
-        public BinaryData(object data, System.Type type, Azure.Core.Serialization.ObjectSerializer? serializer = null) { throw null; }
+        public BinaryData(byte[] data) { throw null; }
+        public BinaryData(object serializable, Azure.Core.Serialization.ObjectSerializer serializer, System.Type? type = null) { throw null; }
+        public BinaryData(object jsonSerializable, System.Type? type = null) { throw null; }
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:10]: ugh

Comment (pakrym) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:10]: This is also unsafe. JsonSerializer always takes type for a reason, sometimes a subclass contains more information than the type you were intending to serialize, not having to pass type might result in a security issue.

File: sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs [hunk 2]
```diff
@@ -5,26 +5,32 @@ namespace Azure.Core
...
+        public BinaryData(System.ReadOnlyMemory<byte> data) { throw null; }
         public BinaryData(System.ReadOnlySpan<byte> data) { throw null; }
         public BinaryData(string data) { throw null; }
-        public System.ReadOnlyMemory<byte> Bytes { get { throw null; } }
```
Comment (kerri-lee) [inline sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:11]: Shouldn't we still have a way to access the binary data as bytes?

File: sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:10]: /cc @KrzysztofCwalina @tg-msft

Comment (tg-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:10]: This was Krzysztof's suggestion.  The feeling was this is a customer API that they call to get data so it's okay to leave the choice up to them while defaulting to the current instance's type.

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs [hunk 3]
```diff
@@ -53,64 +56,120 @@ public BinaryData(ReadOnlySpan<byte> data)
...
+        ///
         /// <returns>A <see cref="BinaryData"/> instance.</returns>
-        public BinaryData(object data, Type type, ObjectSerializer? serializer = default)
+        public BinaryData(object serializable, ObjectSerializer serializer, Type? type = default)
         {
             using var memoryStream = new MemoryStream();
-            serializer ??= new JsonObjectSerializer();
-            serializer.Serialize(memoryStream, data, type, CancellationToken.None);
+            serializer ??= s_jsonSerializer;
+            serializer.Serialize(memoryStream, serializable, type ?? serializable.GetType(), CancellationToken.None);
```
Comment (tg-msft) [inline sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:100]: What if `serializable` is null?  Should it be `type ?? serializable?.GetType() ?? typeof(object)`?

**Review by tg-msft — APPROVED** [review]

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:100]: yikes! I think we can just not allow null objects.

**Pull request merged**